### ISSUE TITLE
Rootfinding sensitive to the area of the contour

### ISSUE THREAD

**Issue by akntnu** [rparini/cxroots#299]
Hi,
I’m using cxroots to find all the roots of a complex function f(z). When I choose a rectangle of size real(-5000,5000), imag(-5000,5000), it gives me 14 roots, each with multiplicity 1. The roots are symmetric about the imaginary axis. That is it has 7 roots of the form x+iy and other 7 roots of the form -x+iy. 
So, when I choose a rectangle of size real(0,5000), imag(-5000,5000), it gives me only 5 roots (two roots with a multiplicity of 2). Which implies, it did not detect two roots compared to the former case. The interesting thing is, if I specify a circle (2500,2500) instead of a rectangle, it gives me all the seven roots. Why is this happening?

Thanks.

**Comment by rparini** [comment]
Hi,
Thanks for reporting this. If two roots are very close to each other compared to the size of the integration contour then it might be failing to distinguish between two close but distinct, simple roots and one root with multiplicity 2.

If you're able to share here a Python script that defines f(z) and the integration contours I can take a closer look when I get some time. This issue may well be something we can improve on.

In the meantime you could also try fiddling with the arguments of the 'roots' method. For example, changing 'int_method' to '"romb"'. Also, if you're able to pass in a definition for the derivative of f(z) to the 'df' argument of the 'roots' method then that should help with general speed and accuracy (without it cxroots approximates the derivative numerically)

**Comment by akntnu** [comment]
Hi,

Please find attached the python script I use to get the roots. In the script, the complex function f(omega) and its derivative f_prime(omega) are defined. All the parameters that are required to compute the function are defined before defining the function. When you run the code, you will get three plots - 1) visualization of the complex function which will give the location of all the seven roots, 2) the roots obtained by using cxroots with a rectangular contour and 3) the roots obtained by using a circle as a contour.

The interesting thing is that cxroots gives all the seven roots when I use the circle instead of a rectangle.

Also, I tried to change the int_method to romb suing the following command,

>> result=contour.roots(f,f_prime,int_method:='romb')

But I was getting the following error.

Error: find_roots() takes from 2 to 3 positional arguments but 4 were given

I shall be grateful to you if you could help me.

Thanking you in advance.

Regards,
abin
________________________________

**Comment by akntnu** [comment]
Hi,

Please find the python script below, which I use to get the roots. In the script, the complex function f(omega) and its derivative f_prime(omega) are defined. All the parameters that are required to compute the function are defined before defining the function. When you run the code, you will get three plots - 1) visualization of the complex function which will give the location of all the seven roots, 2) the roots obtained by using cxroots with a rectangular contour and 3) the roots obtained by using a circle as a contour.

The interesting thing is that cxroots gives all the seven roots when I use the circle instead of a rectangle. 

Also, I tried to change the int_method to romb suing the following command, 

>> result=contour.roots(f,f_prime,int_method:='romb')

But I was getting the following error.

Error: find_roots() takes from 2 to 3 positional arguments but 4 were given

I shall be grateful to you if you could help me.

Thanking you in advance.

Regards,
abin 


`# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""

## Code to compute the frequency and growth rate of thermoacoustic oscillations in a simple 1D geometry##
import numpy as np
import cmath
import mpmath as mp
from mpmath import findroot
from cxroots import Rectangle
from cxroots import Circle
import matplotlib.pyplot as plt
import cplot

#Geometry Details
l=float(1.0)
a=float(0.25)
b=float(l-a)


## Region 1 (unburnt gas)
T1=float(300)
p1=float(101325)
R= 287.0      # Gas constant
rho1=float(p1/(R*T1)) # Density of the unburnt gas (kg/m3)
gamma=1.4
C1=np.sqrt(gamma*R*T1)  #Speed of sound


## Region 2 (burnt gas)
T2=float(1200)
p2=p1
rho2=p2/(R*T2)
C2=np.sqrt(gamma*R*T2)

zeta=float(rho1*C1/(rho2*C2))

##Crocco's n-tau model
n=float(10.0)
tau=float(0.005)

def f(omega):
    return(zeta*cmath.cos(omega/C1*a)*cmath.cos(omega/C2*b)-(1+n*cmath.exp(1j*omega*tau))*(cmath.sin(omega/C1*a)*cmath.sin(omega/C2*b)))

def f_prime(omega):  # Derivative of f
    return(-zeta*(b/C2*cmath.cos(omega*a/C1)*cmath.sin(omega*b/C2)+a/C1*cmath.cos(omega*b/C2)*cmath.sin(omega*a/C1))-(1+n*cmath.exp(1j*omega*tau))*(b/C2*cmath.sin(omega*a/C1)*cmath.cos(omega*b/C2)+a/C1*cmath.sin(omega*b/C2)*cmath.cos(omega*a/C1))-(1j*n*tau*cmath.exp(1j*omega*tau))*(cmath.sin(omega*a/C1)*cmath.sin(omega*b/C2)))

def f1(omega):    # Replacing cmath with np
    return(zeta*np.cos(omega/C1*a)*np.cos(omega/C2*b)-(1+n*np.exp(1j*omega*tau))*(np.sin(omega/C1*a)*np.sin(omega/C2*b)))

#Looking for any real roots
plt=cplot.plot(f1,(0,5000,5000),(-600,600,1200))
plt.show()

print('\nFinding the roots using cxroots\n')

## Using a rectangle as the contour
contour=Rectangle([0,5000],[-500,500])
num_roots = contour.count_roots(f)   # Counting the number of roots within the contour
print('\nThe number of roots within the contour is %d\n'%num_roots)
result=contour.roots(f, f_prime)#,guess_roots_symmetry=lambda omega: [omega.conjugate(), -omega])
#result = contour.roots(f,f_prime)
roots, multiplicities = result
print(result)
result.show()
     
## Using a cirlce as the contour
contour=Circle(2500,2500)
num_roots = contour.count_roots(f)   # Counting the number of roots within the contour
print('\nThe number of roots within the contour is %d\n'%num_roots)
result = contour.roots(f,f_prime)
print(result)
result.show()`

**Comment by rparini** [comment]
Hi,
Thanks for the update. I'm having some trouble reproducing since some of the formatting is ambiguous (ie is `omega/C2b` supposed to be `(omega/C2)*b` or `omega/(C2*b)`?) Would you be able to attach the python file you're using itself (instructions [here](https://docs.github.com/en/github-ae@latest/get-started/writing-on-github/working-with-advanced-formatting/attaching-files)), rather than copy pasting the contents? I should then be able to run exactly the same code you are.

> Also, I tried to change the int_method to romb suing the following command,
> result=contour.roots(f,f_prime,int_method:='romb')
> But I was getting the following error.
> Error: find_roots() takes from 2 to 3 positional arguments but 4 were given

To assign values to function arguments you need to use `=` instead of `:=` so
```
result=contour.roots(f,f_prime,int_method='romb')
```
should work

**Comment by akntnu** [comment]
Hi,

Please find the python code that I use for my work.

I'm sorry for the late reply. 

Looking forward to your response.

Thanking you in advance.

Sincerely yours,
abin


```python
import numpy as np
import cmath
import mpmath as mp
from mpmath import findroot
from cxroots import Rectangle
from cxroots import Circle
import matplotlib.pyplot as plt
import cplot

#Geometry Details
l=float(1.0)
a=float(0.25)
b=float(l-a)

# Region 1 (unburnt gas)
T1=float(300)
P1=float(101325)
R= 287.0      # Gas constant
rho1=float(P1/(R*T1)) # Density of the unburnt gas (kg/m3)
gamma=1.4
C1=np.sqrt(gamma*R*T1)  #Speed of sound

# Region 2 (burnt gas)
T2=float(1200)
P2=P1
rho2=P2/(R*T2)
C2=np.sqrt(gamma*R*T2)

zeta=float(rho1*C1/(rho2*C2))

# Quarter-wave frequency and time period
T=float((4*l)/C1)     # Time period of quarter wave mode in seconds

#Crocco's n-tau model
n=float(10.0)
tau=float(0.005)

def f(omega):
    return(zeta*cmath.cos(omega/C1*a)*cmath.cos(omega/C2*b)-(1+n*cmath.exp(1j*omega*tau))*(cmath.sin(omega/C1*a)*cmath.sin(omega/C2*b)))

def f_prime(omega):  # Derivative of f
    return(-zeta*(b/C2*cmath.cos(omega*a/C1)*cmath.sin(omega*b/C2)+a/C1*cmath.cos(omega*b/C2)*cmath.sin(omega*a/C1))-(1+n*cmath.exp(1j*omega*tau))*(b/C2*cmath.sin(omega*a/C1)*cmath.cos(omega*b/C2)+a/C1*cmath.sin(omega*b/C2)*cmath.cos(omega*a/C1))-(1j*n*tau*cmath.exp(1j*omega*tau))*(cmath.sin(omega*a/C1)*cmath.sin(omega*b/C2)))

def f1(omega):    # Replacing cmath with np
    return(zeta*np.cos(omega/C1*a)*np.cos(omega/C2*b)-(1+n*np.exp(1j*omega*tau))*(np.sin(omega/C1*a)*np.sin(omega/C2*b)))


#Looking for any real roots
fig1, ax1 = plt.subplots()
plt=cplot.plot(f1,(0,5000,5000),(-600,600,1200))
ax1.set_title("n=%0.3f, tau=%0.3f ms"% (n,tau),position=(0.5,1))
plt.show()

# Using a rectangle as the contour
contour1=Rectangle([0,5000], [-500,500])
num_roots1 = contour1.count_roots(f)   # Counting the number of roots within the contour
print('\nThe number of roots within the contour is %d\n'%num_roots1)
result1=contour1.roots(f,f_prime)#,int_method='romb')
roots1, multiplicities1 = result1
print(result1)
result1.show()
 
# Using a cirlce as the contour
contour2=Circle(2500,2500)
num_roots2 = contour2.count_roots(f)   # Counting the number of roots within the contour
print('\nThe number of roots within the contour is %d\n'%num_roots2)
result2 = contour2.roots(f,f_prime)
roots2, multiplicities2 = result2
print(result2)
result2.show()


```

**Comment by rparini** [comment]
Thanks for that code, I can reproduce the issue now.

To give some background: cxroots will subdivide the given integration contour until there are at most `M` (`M`=5 by default) roots (including multiplicities so 1 root with multiplicity 2 counts the same as 2 simple roots)  in a contour, then will use a root approximation algorithm to determine the multiplicities and approximate the roots, which are then further refined by an iterative rootfinding method. 

It looks like in your rectangle case it's finding there's 7 roots (including multiplicites), then subdividing that into two contours with 2 and 5 roots. Then in the 5 root one the approximation algorithm is incorrectly finding some roots with multiplicity 2 and some with multiplicity 0 which is then causing the incorrect result you're seeing. I'll have a look at why that's happening and what a good resolution would be generally.

However, in the meantime, reducing `M` so it subdivides more resolves your particular issue and gives the same answer as the circle
```
result1=contour1.roots(f, f_prime, M=3)
```
Note though that if you do have any roots with multiplicites > 3 then this will make the algorithm very slow.

**Comment by rparini** [comment]
@akntnu thanks for reporting this issue, it should be resolved in the 2.1.0 version of cxroots which has just been released
